src/Internal/Service/Blockfrost.purs

```diff
@@ -253,11 +379,360 @@ getTxMetadata txHash = do
+  { address :: Address
+  , amount :: Value
+  , datum :: OutputDatum
+  , scriptHash :: Maybe ScriptHash
+  }
+
+derive instance Generic BlockfrostTransactionOutput _
+derive instance Newtype BlockfrostTransactionOutput _
+
+instance Show BlockfrostTransactionOutput where
+  show = genericShow
+
+instance DecodeAeson BlockfrostTransactionOutput where
+  decodeAeson = aesonObject \obj -> do
+    address <- decodeAddress obj
+    amount <- decodeValue obj
+    datum <- decodeOutputDatum obj
+    scriptHash <- getFieldOptional' obj "reference_script_hash"
+    pure $ wrap { address, amount, datum, scriptHash }
+    where
+    decodeAddress :: Object Aeson -> Either JsonDecodeError Address
+    decodeAddress obj =
+      getField obj "address" >>= \address ->
+        note (TypeMismatch "Expected bech32 encoded address")
+          (addressFromBech32 address)
+
+    decodeValue :: Object Aeson -> Either JsonDecodeError Value
+    decodeValue =
+      flip getField "amount" >=> aesonArray (map fold <<< traverse decodeAsset)
+      where
+      decodeAsset :: Aeson -> Either JsonDecodeError Value
+      decodeAsset = aesonObject \obj -> do
+        quantity <-
+          getField obj "quantity" >>=
+            BigInt.fromString >>>
+              note (TypeMismatch "Expected string repr of BigInt")
+        getField obj "unit" >>= case _ of
+          "lovelace" -> pure $ Value.lovelaceValueOf quantity
+          assetString -> do
+            let { before: csStr, after: tnStr } = String.splitAt 56 assetString
+            decodeAssetClass assetString csStr tnStr <#> \(cs /\ tn) ->
+              Value.mkValue mempty $ Value.mkSingletonNonAdaAsset cs tn quantity
+
+    decodeOutputDatum :: Object Aeson -> Either JsonDecodeError OutputDatum
+    decodeOutputDatum obj =
+      getFieldOptional' obj "inline_datum" >>= case _ of
+        Just datum ->
+          note (TypeMismatch "Expected CBOR encoded inline datum")
+            (OutputDatum <$> deserializeData datum)
+        Nothing ->
+          maybe NoOutputDatum OutputDatumHash
+            <$> getFieldOptional' obj "data_hash"
+
+resolveBlockfrostTxOutput
+  :: BlockfrostTransactionOutput
+  -> BlockfrostServiceM (Either ClientError TransactionOutput)
+resolveBlockfrostTxOutput
+  (BlockfrostTransactionOutput blockfrostTxOutput@{ address, amount, datum }) =
+  map mkTxOutput <$> resolveScriptRef
+  where
+  mkTxOutput :: Maybe ScriptRef -> TransactionOutput
+  mkTxOutput scriptRef =
+    TransactionOutput { address, amount, datum, scriptRef }
+
+  resolveScriptRef :: BlockfrostServiceM (Either ClientError (Maybe ScriptRef))
+  resolveScriptRef =
+    case blockfrostTxOutput.scriptHash of
+      Nothing -> pure $ Right Nothing
+      Just scriptHash -> runExceptT do
+        scriptRef <- ExceptT $ getScriptByHash scriptHash
+        except $ Just <$> flip note scriptRef
+          (ClientOtherError "Blockfrost: Failed to resolve reference script")
```

Comment:
Can use `for` if you lift the `runExceptT` higher, also `liftMaybe err == except <<< note err` (or something)

```suggestion
  resolveScriptRef = runExceptT do
    for blockfrostTxOutput.scriptHash \scriptHash -> do
      scriptRef <- ExceptT $ getScriptByHash scriptHash
      flip liftMaybe scriptRef
        (ClientOtherError "Blockfrost: Failed to resolve reference script")
```